storage/store.go

```diff
@@ -1075,8 +1086,13 @@ func (s *Store) ProcessRangeDescriptorUpdate(rng *Range) error {
 	}
 	delete(s.uninitRanges, rng.Desc().RaftID)
 
-	s.rangesByKey = append(s.rangesByKey, rng)
-	sort.Sort(s.rangesByKey)
+	if s.rangesByKey.Has((*rangeBTreeItem)(rng)) {
+		return &rangeAlreadyExists{rng}
+	}
+	if exRngItem := s.rangesByKey.ReplaceOrInsert((*rangeBTreeItem)(rng)); exRngItem != nil {
```

Comment:
I think so. We could use `panic` since it never happens, but I wasn't sure how aggressively we should use `panic` to be consistent with the existing code.